File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int treeheight(TreeNode*root){
        if(root==NULL)return 0;
        int lheight=treeheight(root->left);
        if(lheight==-1)return -1;
        int rheight=treeheight(root->right);
        if(rheight==-1)return -1;
        if(abs(lheight-rheight)>1)return -1;
        return max(lheight,rheight)+1;
    }
    bool isBalanced(TreeNode* root) {
        return treeheight(root)!=-1;
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    //logic is similar to finding the height of a bt 
    int check(TreeNode root){
        //check for null node
        if(root==null)return 0;
        int lheight=check(root.left);
        int rheight=check(root.right);
        //if lh || rh at any point are -1 ret -1
        if(lheight==-1 || rheight==-1)return -1;

        //at any pt diff left-right node >1 ret -1
        if(Math.abs(lheight-rheight)>1)return -1;
        
        return 1+Math.max(lheight,rheight);
    }
    public boolean isBalanced(TreeNode root) {
        if(check(root)==-1)return false;
        return true;
    }
}